keps/sig-storage/20190120-execution-hook-design.md

```diff
@@ -0,0 +1,537 @@
+
+        // ActionSucceed is set to true when the action is executed in the container successfully.
+	// It will be set to false if the action cannot be executed successfully after ActionTimeoutSeconds passes.
+        // +optional
+        Succeed *bool
+
+        // The last error encountered when executing the action. The hook controller might update this field each time
+	// it retries the execution.
+	// +optional
+        Error *HookError
+}
+
+type HookError struct {
+        // Type of the error
+        // This is required
+        ErrorType ErrorType
+
+        // Error message
+        // +optional
+        Message *string
+
+        // More detailed reason why error happens
+        // +optional
+        Reason *string
+	
+        // It indicates when the error occurred
+	// +optional
+        Timestamp *int64
+}
+
+type ErrorType string
+
+// More error types could be added, e.g., Forbidden, Unauthorized, AlreadyInProgress, etc.
+const (
+        // The execution hook times out
+        Timeout ErrorType = "Timeout"
+
+        // The execution hook fails with an error
+        Error ErrorType = "Error"
+)
+```
+
+In the ExecutionHookStatus object, there is a list of ContainerExecutionHookStatus for all selected containers in the pods, each ContainerExecutionHookStatus represents the state of the hook on a specific container.
+
+The ActionSucceed field is a summary status field to indicate the final status of the Action in the hook after the commands are run on all selected containers. If not set, this field is nil, meaning the hook controller still needs to wait for the Action commands to finish running on all containers. If set, either true or false, it means the Action commands have finished running (or timeout) on all containers.
+
+
+Here is the definition of HookAction:
+
+```
+// HookAction describes action commands to run on pods/containers based
+// on specified policies. HookAction will be created by the user and
+// can be re-used later. Snapshot Controller will create ExecutionHooks
+// based on HookActions specified in the snapshot spec. For example,
+// two HookActions, preSnapshotExecutionHook and postSnapshotExecutionHook,
+// are expected in the snapshot spec.
+// HookAction does not contain information on pods/containers because those are
+// runtime info.
+// HookAction is namespaced
+type HookAction struct {
+        metav1.TypeMeta
+        // +optional
+        metav1.ObjectMeta
+
+        // This contains the command to run on a container.
+	// The command should be idempotent because the system does not guarantee exactly-once semantics.
+	// Any action may be triggered more than once but only the latest results will be logged in status.
+	// As alpha feature, only ExecAction type in Handler will be support, not the HTTPGETAction or TCPSocketAction.
+        // This is required.
+        Action core_v1.Handler
+
+        // ActionTimeoutSeconds defines when the execution hook controller should stop retrying.
```

Comment:
Should also define the retry policy somewhere: what happens if `ActionTimeoutSeconds` is not specified (i.e. it retries until object is deleted). And is there any exponential backoff policy or is it just back to back retries on failures?

Reply:
Added comments about retry policy.